### DISCUSSION TITLE
[Q&A] glpi-agent - error can't use string ("") as HASH

### DISCUSSION THREAD

**Discussion by Chico0008** [glpi-project/glpi-agent#338] [Q&A]
### Documentation & bug reporting acknowledgment

Yes, I read it

### Describe your problem

Hi

I had a network inventory task running fine since last week.
but now, task seems to not work anymore.

tries running it manually, and in my agent log, got this
`[Wed Feb  8 14:42:08 2023][error] Can't use string ("") as a HASH ref while "strict refs" in use at /usr/share/glpi-agent/lib/GLPI/Agent/Task/NetInventory.pm line 55.`

then, nothing else, but on my server, in glpi inventory / taks, it still see the task as running :/

Nothing changed between the time it worked and now.
tried reinstall the agent : same.
Agent is running on a Ubuntu server, with all modules installed.
```
GLPI Agent (1.5-git75728309)
Built by Debian
Source time: 2023-01-27 02:26
```

**Comment by g-bougard** [comment]
Hi @Chico0008 
the error seems to appear as the server knows nothing about the device to inventory.
Do you always run the netdiscovery task before running the netinventory one ?
Or maybe no device was found in a range ?

**Reply by Chico0008** [reply]
I ran netdiscovery once so i have my printer scanned
then i only ran inventory to the printers
for now it worked fine.

i always have to run a discovery, then a inventory, even if printer are already known in the inventory ?

**Reply by g-bougard** [reply]
My guess is for some reason the plugin lost the required printer information. Maybe a manual updated you don't remember. But in that case, you may see some locked values in the "Locks" tab on the device. Or maybe a GLPI automatic action cleaned the required information.
So try to run again the netdiscovery task to still check this fixes your problem. Then if it is, you may need to run the task at least one or 2 times a week.

**Reply by Chico0008** [reply]
Right
checked : some printer lost their IP address somewhere.
scanned again so they get their IP again, launched inventory manually, OK
i'll see if planned inventory is ok again tonight.

**Reply by Chico0008** [reply]
Worked manually, but then again, not planned
got this in glpiinventory status (same for all the printers)
![image](https://user-images.githubusercontent.com/58886243/218044264-9cf94977-c5b5-403a-bfa8-458d5c36ed1b.png)

the run xxx680 was done manually the morning and ok
the run xxxb59 was planned and KO

on my server, got theses logs
```
[Thu Feb  9 19:13:58 2023][info] target server0: server http://glpi.viry.net/plugins/glpiinventory/
[Thu Feb  9 19:13:58 2023][info] sending contact request to server0
[Thu Feb  9 19:13:58 2023][debug] server message: remoteinventory task not supported
[Thu Feb  9 19:13:58 2023][info] sending prolog request to server0
[Thu Feb  9 19:13:59 2023][error] fork failed:
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 19:13:59 2023][info] target server0: next run: Thu Feb  9 20:59:52 2023 - http://glpi.viry.net/plugins/glpiinventory/
[Thu Feb  9 20:59:52 2023][info] target server0: server http://glpi.viry.net/plugins/glpiinventory/
[Thu Feb  9 20:59:52 2023][info] sending contact request to server0
[Thu Feb  9 20:59:52 2023][debug] server message: remoteinventory task not supported
[Thu Feb  9 20:59:52 2023][info] sending prolog request to server0
[Thu Feb  9 20:59:52 2023][error] fork failed:
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][error] fork failed: Inappropriate ioctl for device
[Thu Feb  9 20:59:52 2023][info] target server0: next run: Thu Feb  9 22:43:53 2023 - http://glpi.viry.net/plugins/glpiinventory/
```

my glpi-agent is installed and run as root, with log configuration to file > /var/log/glpi-agent.log
file is root:root 644
here's my log conf
```
# Logger backend, either Stderr, File or Syslog (Stderr)
logger = file
# log file
logfile = /var/log/glpi-agent.log
# maximum log file size, in MB
logfile-maxsize = 8
# Syslog facility
logfacility = LOG_USER
# Use color in the console
color = 0
#
# Execution mode options
#
# add given tag to inventory results
tag =
# debug mode
debug = 2
```

and i have a glpi-agent process running NetInventory stuck

**Reply by g-bougard** [reply]
`fork failed: ...`... wtf, didn't see such message before.
You probably have messed up something. Have you cleanly restarted the service ? Try to stop the service, eventually kill remaining glpi-agent processes and restart it again. Then try again to force your task.

**Reply by Chico0008** [reply]
Yep, service has been restarted fine

i modify the task to had it run / re-preper always (no prep widows nor run windows)
it runned fine several times.

i setup my task again with prep windows and run windows, i'll see on monday what happens.
